apollo-router/src/query_planner/plan.rs

```diff
@@ -188,22 +188,45 @@ impl PlanNode {
         }
     }
 
-    pub(crate) fn subgraph_fetches(&self) -> usize {
+    /// Count the number of fetches
+    ///
+    pub(crate) fn subgraph_fetches(&self, include_requires: bool) -> usize {
```

Comment:
i suspect that counting fetches is not going to be enough.
consider this use case: 

![image](https://github.com/apollographql/router/assets/6611331/1a6090a5-d08c-47f7-8618-7087cb084106)

the query 
```graphql
query Astronaut {
  id
  name
  missions {
    designation
  }
} 
```
The query plan of above query consists in:
a SequenceNode with 2 children: a FetchNode (astronauts subgraph) and a FlattenNode,
the FlattenNode with a FetchNode (missionsSubgraph).

if we "count" the fetches, we have a grand total of 2, however, fetch #2 will never fire if fetch #1 does not complete first.

now imagine we send a batch of 2 graphqlRequests using above query.
the # of fetches will be 4, while realistically we won't be able to fire fetch #3 and fetch #4 before #1 and #2 completes.

IMHO this is only solvable with dataloaders and stop using async/await in query plan executor

Reply:
@samuelAndalon Thanks for the feedback!

I wrote about this limitation (and the assumptions) in Section 2.1.1 of the shared design documentation and it is one of the major assumptions that we aim to test in the prototype.

I believe the prototype accounts for that by excluding fetches which have "required" fields from the total.

I'm finding this works fine in the prototyping graphs/queries I'm working with, but I'm not 100% sure it will work for all use cases and that's one of the things which will require more testing.

We'll need to move onto alternative approaches (or further refinements) if this assumption is proved invalid during prototyping.